### PR TITLE
[xcbuild] Check ~/.xcsdk/xcode_dir_path when getting developer root

### PR DESCRIPTION
**Pull request by Ktwu**
xcrun attempts to determine the developer root so that it knows where
how to locate toolchains and sdks.
Currently the developer root is resolved through the following steps
1. Checking DEVELOPER_DIR environment variable
2. Check the /var/db/xcode_select_link
3. /usr/share/xcode-select/xcode_dir_path
(1) suffers from having to configure this env var everytime xcrun is run.
(2) and (3) suffers on needing to be root to configure on most systems.

This change adds another symlink to check for:
~/.xcsdk/xcode_dir_path

Test Plan:
1. Create symlink ~/.xcsdk/xcode_dir_path
2. xcrun -v -l --sdk test-sdk -f test-script

Change-Id: I88ed516e7264cffa263f6ed831e34083930e3337

### CODE REVIEW

**Comment by Ktwu** [comment]
PR on Naris' behalf.

**Comment by grp** [comment]
FWIW `xcode_select_link` is the newer name, so maybe match that for this rather than `xcode_dir_path`?

File: Libraries/xcsdk/Sources/Environment.cpp [hunk 1]
```diff
@@ -17,6 +21,16 @@ using libutil::Filesystem;
 using libutil::FSUtil;
 
 static std::string
+UserDeveloperRootLink()
+{
+    char *home_dir = getenv("HOME");
+    if (home_dir == NULL) {
+        home_dir = getpwuid(getuid())->pw_dir;
+    }
+    return std::string(home_dir) + "/.xcsdk/xcode_dir_path";
```
Comment (grp) [inline Libraries/xcsdk/Sources/Environment.cpp]: Ah, the other place in this file uses `getenv("HOME")`.

**Comment by Ktwu** [comment]
When you say `xcode_select_link` is the newer name, does this mean `xcode_dir_path` is no longer supported?

**Comment by grp** [comment]
It's still the fallback, but the other one is preferred if available.

Sent from my iPhone

On Sep 20, 2016, at 02:10, Ktwu <notifications@github.com> wrote:

When you say xcode_select_link is the newer name, does this mean
xcode_dir_path is no longer supported?

—
You are receiving this because you were assigned.
Reply to this email directly, view it on GitHub
<https://github.com/facebook/xcbuild/pull/121#issuecomment-248092128>, or mute
the thread

**Comment by Ktwu** [comment]
Ok, updated the path to xcode_select_link.

**Comment by grp** [comment]
How soon do you need this in? I think it might be easier to rebase this on #124 than the reverse.

**Comment by Ktwu** [comment]
This doesn't need to go in soon -- my team has local changes that we're working with.

**Comment by Ktwu** [comment]
Just rebased.

File: Libraries/xcsdk/Sources/Environment.cpp [hunk 2]
```diff
@@ -18,6 +22,16 @@ using libutil::Filesystem;
 using libutil::FSUtil;
 
 static std::string
+UserDeveloperRootLink()
+{
+    char *home_dir = getenv("HOME");
+    if (home_dir == NULL) {
+        home_dir = getpwuid(getuid())->pw_dir;
+    }
```
Comment (grp) [inline Libraries/xcsdk/Sources/Environment.cpp]: There's a method on `process::Context` for this now.

**Comment by grp** [comment]
Landed with modifications to use `process::Context`.

**Comment by grp** [comment]
Landed with modifications to use `process::Context`.

**Pull request closed without merging**